### PR TITLE
MODLD-583: LCCN Validation of Pattern

### PR DESCRIPTION
**Pull request by askhat-abishev**

### CODE REVIEW

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  private boolean isCurrent(LccnRequest lccnRequest) {
+    return lccnRequest.getStatus()
+      .stream()
+      .flatMap(status -> status.getLink().stream())
+      .anyMatch(link -> link.endsWith("current"));
+  }
+
+  private boolean hasInvalidPattern(String lccnValue) {
+    return !LCCN_STRUCTURE_A_PATTERN.matcher(lccnValue).matches()
+      && !LCCN_STRUCTURE_B_PATTERN.matcher(lccnValue).matches();
```
Comment (AndreiBordak) [inline src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java:34]: minor: i'd make it positive, mean
```
hasValidPattern(String lccnValue) {
LCCN_STRUCTURE_A_PATTERN.matcher(lccnValue).matches()
      && LCCN_STRUCTURE_B_PATTERN.matcher(lccnValue).matches();
}
```

**Review by pkjacob — APPROVED** [review]

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      return lccnRequest.getValue()
+        .stream()
+        .noneMatch(this::hasInvalidPattern);
+    } else {
+      return true;
+    }
+  }
+
+  private boolean isCurrent(LccnRequest lccnRequest) {
+    return lccnRequest.getStatus()
```
Comment (pkjacob) [inline src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java:26]: Hi @askhat-abishev,
When user do not choose any status in the UI (refer image), then also we have to consider the LCCN as current.

So, we will have to change this logic as follows

```
isEmpty(lccnRequest.getStatus()) || ... the current logic
```

Screenshot 2024-11-18 at 11 31 12 AM

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java (hunk 1 above)
Comment (askhat-abishev) [reply src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java]: done

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java (hunk 2 above)
Comment (askhat-abishev) [reply src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java]: done

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  private static final Pattern LCCN_STRUCTURE_A_PATTERN = Pattern.compile("( {3}|[a-z][ |a-z]{2})\\d{8} ");
+  private static final Pattern LCCN_STRUCTURE_B_PATTERN = Pattern.compile("( {2}|[a-z][ |a-z])\\d{10}");
+
+  @Override
+  public boolean isValid(LccnRequest lccnRequest, ConstraintValidatorContext constraintValidatorContext) {
+    if (isCurrent(lccnRequest)) {
+      return lccnRequest.getValue()
+        .stream()
+        .anyMatch(this::hasValidPattern);
```
Comment (AndreiBordak) [inline src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java:21]: it was noneMatch(this::hasInvalidPattern), is anyMatch(this::hasValidPattern) expected here?

**Review by AndreiBordak — APPROVED** [review]

File: src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java (hunk 3 above)
Comment (askhat-abishev) [reply src/main/java/org/folio/linked/data/validation/dto/LccnPatternValidator.java:21]: sure

**Review by PBobylev — APPROVED** [review]

**Pull request merged**